src/testing/convection_diffusion_explicit_periodic.cpp

```diff
@@ -0,0 +1,443 @@
+        for(int idim=0; idim<dim; idim++){
+            dealii::LinearAlgebra::distributed::Vector<double> Mq_hat(dg->right_hand_side);
+            dg->global_mass_matrix_auxiliary[idim].vmult( Mq_hat, dg->auxiliary_solution[idim]);
+            double temp_energy = dg->auxiliary_solution[idim] * Mq_hat;
+            energy += diff_coeff * temp_energy;
+        }
+    
+	return energy;
+}
+
+template<int dim, int nstate>
+double ConvectionDiffusionPeriodic<dim, nstate>::compute_conservation(std::shared_ptr < PHiLiP::DGBase<dim, double> > &dg, const double poly_degree) const
+{
+	double conservation = 0.0;
+       // dealii::LinearAlgebra::distributed::Vector<double> Mu_hat(dg->solution.size());
+        dealii::LinearAlgebra::distributed::Vector<double> Mu_hat(dg->right_hand_side);
+       // printf(" size Mass %d size sol %d\n",dg->global_mass_matrix.n(),dg->solution.size());
+       // printf("  Mass %g size sol %g\n",dg->global_mass_matrix(dg->global_mass_matrix.m(),dg->global_mass_matrix.n()),dg->solution(dg->solution.size()));
+        //dg->global_mass_matrix.vmult( Mu_hat, dg->right_hand_side);
+        dg->global_mass_matrix.vmult( Mu_hat, dg->solution);
+
+        const unsigned int n_dofs_cell = dg->fe_collection[poly_degree].dofs_per_cell;
+        const unsigned int n_quad_pts = dg->volume_quadrature_collection[poly_degree].size();
+        dealii::Vector<double> ones(n_quad_pts);
+        for(unsigned int iquad=0; iquad<n_quad_pts; iquad++){
+            ones[iquad] = 1.0;
+        }
+        dealii::Vector<double> ones_hat(n_dofs_cell);
+        dg->operators.vol_projection_operator[poly_degree].vmult(ones_hat, ones);
+
+        dealii::LinearAlgebra::distributed::Vector<double> ones_hat_global(dg->right_hand_side);
+        std::vector<dealii::types::global_dof_index> dofs_indices (n_dofs_cell);
+        for (auto cell = dg->dof_handler.begin_active(); cell!=dg->dof_handler.end(); ++cell) {
+            if (!cell->is_locally_owned()) continue;
+            cell->get_dof_indices (dofs_indices);
+            for(unsigned int idof=0;idof<n_dofs_cell; idof++){
+                ones_hat_global[dofs_indices[idof]] = ones_hat[idof];
+            }
+        }
+
+        conservation = ones_hat_global * Mu_hat;
+
+        const double diff_coeff = Parameters::ManufacturedSolutionParam::get_default_diffusion_coefficient();
+        for(int idim=0; idim<dim; idim++){
+            dealii::LinearAlgebra::distributed::Vector<double> Mq_hat(dg->right_hand_side);
+            dg->global_mass_matrix_auxiliary[idim].vmult( Mq_hat, dg->auxiliary_solution[idim]);
+            double temp_conservation = ones_hat_global * Mq_hat;
+            conservation += diff_coeff * temp_conservation;
+        }
+
+	return conservation;
+}
+
+
+template<int dim, int nstate>
+void ConvectionDiffusionPeriodic<dim, nstate>::initialize(PHiLiP::DGBase<dim, double>  &dg,
```

Comment:
you could move this to the `InitialConditionFunction` class in `src/physics/initial_condition/` and make use of the factory

Reply:
fixed as per issue #123